src/BaseControllerV2.ts

```diff
@@ -9,6 +9,15 @@ import type { Draft } from 'immer';
  */
 export type Listener<T> = (state: T) => void;
 
+export type Anonymizer<T> = (value: T) => T;
```

Comment:
I'm not sure whether this is typed correctly. For example, let's say `T` is some interface, and the anonymizer deletes some keys on the object that implements `T`. In that case, I believe that the return value won't be `T`, and there will be much complaining from TypeScript.

I think this would be better:
```suggestion
export type Anonymizer<T, U> = (value: T) => U;
```

Reply:
Good point. I had only tested this in cases where the anonymized state is the same type.

Reply:
Adding a second generic parameter doesn't really work though. What would we set it to?

Maybe something like `T | Partial<T>` would work, to at least let us omit properties. We'd have to preserve types for any properties that remain, but that might be OK.

Reply:
> Adding a second generic parameter doesn't really work though. What would we set it to?

It can either be set to `Partial<T>` by the caller (defaulting to that would be nice, but I don't know if that's possible), or I imagine something like this:

```typescript
interface MyControllerState { ... }
interface MyControllerAnonymousState { ... }
```

The simplest way to define the `AnonymousState` interface is probably by using the [`Omit` utility type](https://www.typescriptlang.org/docs/handbook/utility-types.html#omittype-keys) on `MyControllerState`, e.g. `type MyControllerAnonymousState = Omit<MyControllerState, ...keys>`.

If we add a second generic to `Anonymizer`, it would have to propagate to all of its consumers, so we'd ultimately end up with:

```typescript
class MyController extends BaseController<MyControllerState, MyControllerAnonymousState> { ... }
```

Reply:
Huh, that's an interesting idea! I'm not sure it's worth the effort at this point though, if we can find a default that works for all cases we can think of right now. I've been working on a recursive partial type that might be good enough for anything we'd want to do.

This is something we should keep in mind if we ever feel the need to work more with the anonymized state and want better typing, or if we want an anonymizer to return a totally different type at any point. It looks like TypeScript does support default generic parameters, so we might be able to do this in a way that doesn't burden every controller with having to declare this.

Reply:
Here's what I have so far: 3f41181f25055d0a0d7a5f0e645e84399a3734c7

Let me know what you think!